core/string/ustring.cpp

```diff
@@ -4545,84 +4545,128 @@ bool String::is_network_share_path() const {
 	return begins_with("//") || begins_with("\\\\");
 }
 
-String String::simplify_path() const {
-	String s = *this;
-	String drive;
-
+static int find_drive_end_idx(const String &p_str) {
 	// Check if we have a special path (like res://) or a protocol identifier.
-	int p = s.find("://");
-	bool found = false;
+	int p = p_str.find("://");
 	if (p > 0) {
 		bool only_chars = true;
 		for (int i = 0; i < p; i++) {
-			if (!is_ascii_alphanumeric_char(s[i])) {
+			if (!is_ascii_alphanumeric_char(p_str[i])) {
 				only_chars = false;
 				break;
 			}
 		}
 		if (only_chars) {
-			found = true;
-			drive = s.substr(0, p + 3);
-			s = s.substr(p + 3);
-		}
-	}
-	if (!found) {
-		if (is_network_share_path()) {
-			// Network path, beginning with // or \\.
-			drive = s.substr(0, 2);
-			s = s.substr(2);
-		} else if (s.begins_with("/") || s.begins_with("\\")) {
-			// Absolute path.
-			drive = s.substr(0, 1);
-			s = s.substr(1);
-		} else {
-			// Windows-style drive path, like C:/ or C:\.
-			p = s.find(":/");
-			if (p == -1) {
-				p = s.find(":\\");
-			}
-			if (p != -1 && p < s.find_char('/')) {
-				drive = s.substr(0, p + 2);
-				s = s.substr(p + 2);
-			}
+			return p + 3;
 		}
 	}
 
-	s = s.replace("\\", "/");
-	while (true) { // in case of using 2 or more slash
-		String compare = s.replace("//", "/");
-		if (s == compare) {
-			break;
-		} else {
-			s = compare;
-		}
+	if (p_str.is_network_share_path()) {
+		// Network path, beginning with // or \\.
+		return 2;
+	}
+
+	if (p_str.begins_with("/") || p_str.begins_with("\\")) {
+		// Absolute path.
+		return 1;
 	}
-	Vector<String> dirs = s.split("/", false);
 
-	for (int i = 0; i < dirs.size(); i++) {
-		String d = dirs[i];
-		if (d == ".") {
-			dirs.remove_at(i);
-			i--;
-		} else if (d == "..") {
-			if (i != 0) {
-				dirs.remove_at(i);
-				dirs.remove_at(i - 1);
-				i -= 2;
+	// Windows-style drive path, like C:/ or C:\.
+	p = p_str.find(":/");
+	if (p == -1) {
+		p = p_str.find(":\\");
+	}
+	if (p != -1) {
+		bool no_separator = true;
+		const char32_t *ptr = p_str.ptr();
+		for (const char32_t *end = ptr + p; ptr < end; ptr++) {
+			if (is_path_seperator(*ptr)) {
+				no_separator = false;
+				break;
 			}
 		}
+		if (no_separator) {
+			return p + 2;
+		}
 	}
 
-	s = "";
+	return 0;
```

Comment:
This is pretty nice to read. But there is probably a version with a simple linear search that could work. Not sure its worth investigating though.

Reply:
What do you mean by 'linear search'? I do regret that my implementation always copies the resulting string, even if no adjustments are needed, but such was the case with the old implementation too.

Reply:
I mean you are doing multiple find calls. There is _probably_ a way to resolve this in one for loop

Reply:
I had it implemented with a 'part stack's previously because I wanted to avoid backtracking, but the overhead from allocation turned out to be way slower than the overhead from backtracking from most normal length strings - especially since `..` is a rare case.

I think there is a theoretical implementation of this function that uses two passes instead of a stack or backtracking - one forward pass to track the amount of `..` (and whether changes are even needed), and another to apply the changes. But it would use a code (and time) to implement.